### DISCUSSION TITLE
Use a pattern'ed URL to fetch the latest version.

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [ImranR98/Obtainium#1034] [Q&A] [answered]
Hello there!

I would like to add RetroArch to Obtanium - but don't really know what the best way would be. 

Their releases can be found here: https://buildbot.libretro.com/stable/

Technically, the URL pattern would be `https://buildbot.libretro.com/stable/{VERSION}/android/RetroArch_aarch64.apk`
But I don't know how I can make Obtanium query the versions found in the `stable` folder, as it is rendered in HTML. And the JS query for it isn't that nice either: `Object.values(document.querySelectorAll("#items li.item.folder a span.label")).map(e => e.innerText)`

Do you have an idea how I could make that work? Their F-Droids builds basically don't exist.

Thanks and kind regards,
Ingwie

**Comment by malijaffri** [comment, marked as answer]
I was able to add RetroArch to Obtainium using the following settings:

| Field | Value |
| ----- | ----- |
| App Source URL | `https://retroarch.com/?page=platforms` |
| APK Link Filter | `buildbot\.libretro\.com/stable/((:?\d+\.)*\d+)/android/.*\.apk$` |
| Version Extraction RegEx | `.*/stable/((:?\d+\.)*\d+)/android/.*` |
| Match Group for Version Extraction RegEx | `1` |
| Filter APKs by RegEx | `\bRetroArch_aarch64.apk$` |

The settings that I have not mentioned I left at the default values.

Let me know if it doesn't work for you.

**Reply by IngwiePhoenix** [reply]
Ok - now all I have to do is figure out how to copy all the RegExes into the app... because I am definitively not copytyping this by hand. I am visually impaired, errors are a guarantee with this xD

Thanks a lot, I will see if I can make scrcpy work and see if I get some form of clipboard sharing working... this should do the trick.

**Reply by IngwiePhoenix** [reply]
Yay for copypasta, "Note to self" channel in Matrix, and the fact that a Pixel Fold can run two things on one screen. Phew, got this done, and it worked! I did, however, get a message about the origin being `retroarch.com` but the actual source being `buildbot.libretro.com`. But aside from that, it pulled the correct version immediately and installed just fine.

Thanks for your help!

**Reply by malijaffri** [reply]
You're welcome :)

Just as a side note, Obtainium will return the whole message about origin being retroarch.com and source being buildbot.libretro.com every time you update the app. It's not an error, just a warning in case something messed up during the configuration or something changed in the webpage structure itself later on.

**Comment by davidhedlund** [comment]
Can you guys please take a look at:
* https://github.com/ImranR98/apps.obtainium.imranr.dev/issues/1170
